src/command/list/toolchain.rs

```diff
@@ -0,0 +1,345 @@
+
+        let package_managers = inventory
+            .yarn
+            .versions
+            .iter()
+            .map(|version| PackageManager {
+                kind: PackageManagerKind::Yarn,
+                source: Lookup::Yarn.version_source(project, user_platform, version),
+                version: version.clone(),
+            })
+            .collect();
+
+        let packages = Package::from_inventory_and_project(inventory, project);
+
+        Ok(Toolchain::All {
+            runtimes,
+            package_managers,
+            packages,
+        })
+    }
+
+    pub(super) fn node(
+        inventory: &Inventory,
+        project: &Option<Rc<Project>>,
+        user_platform: &Option<Rc<PlatformSpec>>,
+        filter: &Filter,
+    ) -> Toolchain {
+        let runtimes = inventory
+            .node
+            .versions
+            .iter()
+            .filter_map(|version| {
+                let source = Lookup::Runtime.version_source(project, user_platform, version);
+                if source.allowed_with(filter) {
+                    let version = version.clone();
+                    Some(Node { source, version })
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        Toolchain::Node(runtimes)
+    }
+
+    pub(super) fn yarn(
+        inventory: &Inventory,
+        project: &Option<Rc<Project>>,
+        user_platform: &Option<Rc<PlatformSpec>>,
+        filter: &Filter,
+    ) -> Toolchain {
+        let yarns = inventory
+            .yarn
+            .versions
+            .iter()
+            .filter_map(|version| {
+                let source = Lookup::Yarn.version_source(project, user_platform, version);
+                if source.allowed_with(filter) {
+                    Some(PackageManager {
+                        kind: PackageManagerKind::Yarn,
+                        source,
+                        version: version.clone(),
+                    })
+                } else {
+                    None
+                }
+            })
+            .collect();
+
+        Toolchain::PackageManagers(yarns)
+    }
+
+    pub(super) fn package_or_tool(
+        name: &str,
+        inventory: &Inventory,
+        project: &Option<Rc<Project>>,
+        filter: &Filter,
+    ) -> Fallible<Toolchain> {
+        /// An internal-only helper for tracking whether we found a given item
+        /// from the `PackageCollection` as a *package* or as a *tool*.
+        #[derive(PartialEq, Debug)]
+        enum Kind {
+            Package,
+            Tool,
+        }
+
+        /// A convenient name for this tuple, since we have to name it in a few
+        /// spots below.
+        type Triple = (Kind, PackageConfig, Source);
+
+        let (packages_and_tools, errors): (Vec<Fallible<Triple>>, Vec<Fallible<Triple>>) =
+            inventory
+                .packages
+                .clone()
+                .into_iter()
```

Comment:
Similar comment here about potentially providing an `iter` method that returns an Iterator of references.

Reply:
Done!